### DISCUSSION TITLE
Unknown service

### DISCUSSION THREAD

**Discussion by purashy** [outline/outline#3984] [Self Hosting] [answered]
Not sure why but I keep getting this error message and I am not sure why.
Any help is appreciated.

```/outline/build/server/index.js:100
      throw new Error(`Unknown service ${name}`);
            ^

Error: Unknown service 
    at start (/root/outline/build/server/index.js:100:13)
    at throng (/root/outline/node_modules/throng/lib/throng.js:24:18)
    at Object.<anonymous> (/root/outline/build/server/index.js:129:21)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
    at node:internal/main/run_main_module:17:47
```

**Comment by vgwidt** [comment, marked as answer]
Probably has to due with this commit made today: https://github.com/outline/outline/commit/41d7cc26b5b045a4ede5670bf48a06367c69d1e1

Mine is broken too after rebuilding this morning but can't troubleshoot quite yet.

EDIT: reverting to the commit before the one listed above will fix this issue.  I reverted to https://github.com/outline/outline/commit/e57941732a121e30fdf8824bb0214619f333204a

I still don't know why it broke though.  How the services are enumareted was refactored but it looks to be doing the same exact thing.  Some extra debugging at that stage should provide some hints.

**Reply by apoorv-mishra** [reply]
How are you running? With or without the `--services` flag?

**Reply by vgwidt** [reply]
> How are you running? With or without the `--services` flag?

It wont run either way, but the issue is with the use of ?? instead of || for getArgs.  I'll post a PR that fixes it in a sec.

**Reply by apoorv-mishra** [reply]
Exactly! But I'm not sure why it's not running with `--services=<at-least-one-service>` for you. It should.

**Reply by tommoor** [reply]
If you're putting in a PR change `getArg` to return `undefined` when the arg doesn't exist instead. It's a weird interface that it always returns a string imo, hence the mistake.

**Reply by vgwidt** [reply]
Yeah it looks to be due to getArg returning an emptry string.  For now I was just going to change it to:
  public SERVICES =
    getArg("services") ||
    (process.env.SERVICES ??
    "collaboration,websockets,worker,web");

If you guys can make it for me instead (especially if fixing getArg instead) it would probably be a good idea, I have almost no programming experience :(

**Reply by tommoor** [reply]
I'll fix it.

**Reply by apoorv-mishra** [reply]
Just a note. `process.env.SERVICES` would still be interpreted at `""` if missing in `.env`.

**Reply by vgwidt** [reply]
> Just a note. `process.env.SERVICES` would still be interpreted at `""` if missing in `.env`.

I don't have it set in .env and it looks to default to "collaboration,websockets,worker,web" with ?? between the two in my case.

Also, thanks Tom!

**Reply by vgwidt** [reply]
@apoorv-mishra On the note of --services flag not working when testing, I got an error for unknown flag, but I didn't try to troubleshoot it further since I discovered the operator problem.  But it was probably my own mistake.  I'll try again and confirm it when I have some time.

**Reply by purashy** [reply]
Just woke up to this awesome news :) Thanks guys, glad it was quickly found what the issue is. Also thank you @tommoor. eager to try the app out.

**Reply by tommoor** [reply]
I recommend using tagged releases, ideally the official Docker builds – not `main` which can be unstable btw.
